### PR TITLE
[PyROOT][6467] For Python-C++ derived types, invoke constructor via the dispatcher

### PR DESCRIPTION
**Pull request by etejedor**
Fixes https://github.com/root-project/root/issues/6467.

It needs https://github.com/root-project/root/pull/6377 for Python <= 3.6, otherwise an error of the type `KeyError: 'fProofChain'` (for TChain) happens first and we don't even get to the issue that this PR fixes (UPDATE: #6377 has been merged).

Also, before merging, this fix needs to be tested for `TGMainFrame`, where the problem described in https://github.com/root-project/root/issues/6578 has been identified.

### CODE REVIEW

**Comment by etejedor** [comment]
@phsft-bot build

File: bindings/pyroot/cppyy/CPyCppyy/src/CPPConstructor.cxx [hunk 1]
```diff
@@ -41,75 +41,64 @@ PyObject* CPyCppyy::CPPConstructor::Call(
...
-                return nullptr;
-            }
+        PyObject* pyobj = PyObject_Call(dispproxy, args, kwds);
+        if (!pyobj)
+            return nullptr;
 
-            PyObject *pyoff = PyObject_CallMethod(dispproxy, (char*)"_dispatchptr_offset", nullptr);
+    // retrieve the actual pointer, take over control, and set m_self
+        address = (ptrdiff_t)((CPPInstance*)pyobj)->GetObject();
+        if (address) {
```
Comment (stwunsch) [inline bindings/pyroot/cppyy/CPyCppyy/src/CPPConstructor.cxx:84]: What happens if `address` is null?

Comment (stwunsch) [reply bindings/pyroot/cppyy/CPyCppyy/src/CPPConstructor.cxx:84]: The logic is super complicated :sweat_smile:

**Review by stwunsch — APPROVED** [review]
LGTM, in principle, though it's very hard to know from the code whether everything is fine. Would it be possible to have a test or is there no changed behaviour?

File: bindings/pyroot/cppyy/CPyCppyy/src/CPPConstructor.cxx (hunk 1 above)
Comment (etejedor) [reply bindings/pyroot/cppyy/CPyCppyy/src/CPPConstructor.cxx:84]: You will end up setting an error of `constructor failed` further down, in line 136.

**Comment by etejedor** [comment]
> LGTM, in principle, though it's very hard to know from the code whether everything is fine. Would it be possible to have a test or is there no changed behaviour?

I understand the review is a bit hard if you are not familiar with the code. In the commit message I try to explain the reason of the change. As for the test, I added this:

https://github.com/root-project/roottest/pull/607

which reproduces what the commit is trying to fix (a proper match and invocation if the constructor has default arguments). The other tests seem to run fine. I also want to add a test for TGMainFrame once https://github.com/root-project/root/issues/6578 is fixed.

**Comment by stwunsch** [comment]
Do I understand correctly that this is already in upstream cppyy?

**Comment by stwunsch** [comment]
But anyway, seems fine. Just hard to tell whether it breaks something, it's highly non trivial!

**Pull request merged**